crypto/provider_core.c

```diff
@@ -208,7 +208,7 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
         int ref = 0;
 
 #ifndef HAVE_ATOMICS
-        CRYPTO_DOWN_REF(&prov->refcnt, &ref, provider_lock);
+        CRYPTO_DOWN_REF(&prov->refcnt, &ref, prov->refcnt_lock);
```

Comment:
refcnt_lock appears to be NULL?

Reply:
Argh...

The refcount header doesn't make it easy to build without atomics :confused: 
That needs to change.

Reply:
Fixed

Reply:
I do wonder why we're using `HAVE_ATOMICS` here at all. We could just create the lock unconditionally. If HAVE_ATOMICS is defined then it won't be used by the `CRYPTO_DOWN_REF` implementation...but so what?

Reply:
'til some code sanitizer notices and starts complaining about it

Reply:
It's what we do *everywhere* else. This is the only file in the whole repo that actually *uses* the HAVE_ATOMICS define.

Reply:
Okie.  I've now ripped out most of those ifdefs.  I will *refuse* to remove those around `CRYPTO_THREAD_lock_new` and `CRYPTO_THREAD_lock_free`, though.  That would be stoopid waste of resources (and if we *are* wasteful elsewhere, I think guards should be *added*)